lib/tasks/migrate_notes.rake

```diff
@@ -0,0 +1,24 @@
+namespace :db do
+  desc "Backfills notes-columns `body`, `author_ip` and `author_id`"
+  task :migrate_notes => :environment do
+    scope = Note.where(:body => nil, :author => nil, :author_ip => nil)
+    total_count = scope.count
+    remaining_count = total_count
+    puts "A total of #{total_count} Note-records have to be migrated."
+
+    # NB: default batch size is 1000
+    scope.find_in_batches do |batch|
+      puts "Processing batch of #{batch.size} records."
+      batch.each do |record|
+        opened_comment = record.comments.unscope(:where => :visible).find_by(:event => "opened")
+        (putc "x" && next) unless opened_comment
+
+        attributes = opened_comment.attributes.slice(*%w[body author_id author_ip]).compact_blank
+        record.update_columns(attributes) # rubocop:disable Rails/SkipsModelValidations
```

Comment:
I think this needs to remote the comment too, right? Otherwise there will be a body, and then the first comment will duplicate the body.

I also think this needs to have tests, and so I would move all the logic into the app (e.g. Note.migrate_bodies, or a standalone class in lib, or somesuch) to make it easier to test. The rake task can then just be a one-liner to call that method and therefore doesn't need tests itself.

Reply:
> I think this needs to remove the comment too, right? Otherwise there will be a body, and then the first comment will duplicate the body.

I am hesitant to perform any destructive operations until the very last step, but I open to change my approach here since it would simplify the code.

`Note`-model methods (`#body`, `#author`, `#author_ip`) have been adjusted so that calls to these return the migrated data once its present and otherwise look it up through the `opened`-NoteComment-record.

```rb
class Note < ApplicationRecord
  # FIXME: notes_refactoring remove this once the backfilling is completed
  def body_migrated?
    attributes["body"].present?
  end

  # FIXME: notes_refactoring remove this once the backfilling is completed
  def author
    return super if body_migrated?

    opened_note_comment&.author
  end

  # FIXME: notes_refactoring remove this once the backfilling is completed
  def author_ip
    return super if body_migrated?

    opened_note_comment&.author_ip
  end

  # Return the note body
  def body
    body = body_migrated? ? super : opened_note_comment&.body&.to_s
    RichText.new("text", body)
  end
end
```

> I also think this needs to have tests, and so I would move all the logic into the app (e.g. Note.migrate_bodies, or a standalone class in lib, or somesuch) to make it easier to test.

👍🏽 will add this.

Reply:
> I am hesitant to perform any destructive operations until the very last step, but I am open to change the approach here since it would simplify the code.

I'm not sure I understand how this migration works, please let me know if I've missed something obvious! Let's say we start with a note with two comments:

|Situation | Note id | Note body | first comment body | second comment body | 
|---|---|---|---|---|
| A | 123 | - | shop here | I fixed it |
| B | 123 | shop here | shop here | I fixed it | 
| C | 123 | shop here | I fixed it | - |

Situation A is before we do anything - no body, two comments. At the end of the project, we're aiming for situation C.

However, if we migrate the contents of the first comment, without destroying that comment, don't we end up in situation B? i.e. `@note.body_migrated?` is `true`, `build_comments_with_extra_open_comment` will then have three comments (two real ones plus the unshifted synthetic one)?

Here's a test to show my point:

```ruby
  def test_comments_after_migration
    note = create(:note, :body => nil, :author => nil, :author_ip => nil)
    create(:note_comment, :note => note, :event => "opened", :body => "Hey hey!", :author => create(:user), :author_ip => "10.0.0.1")
    create(:note_comment, :note => note, :event => "commented", :body => "done", :author => create(:user), :author_ip => "10.0.0.2")

    assert_equal 2, note.comments_with_extra_open_comment.length

    assert Note::MigrateOpenedComment.new(note).call

    n = Note.find(note.id) # avoids all caching, `||=` caches etc
    assert_equal 2, n.comments_with_extra_open_comment.length
  end
```

```
Failure:
NoteMigrateOpenedCommentTest#test_comments_after_migration [test/models/note/migrate_opened_comment_test.rb:24]:
Expected: 2
  Actual: 3
```